zebra-state/src/service/check/anchors.rs

```diff
@@ -3,107 +3,84 @@
-        // Sprout JoinSplits, with interstitial treestates to check as well.
-        if transaction.has_sprout_joinsplit_data() {
-            let mut interstitial_trees: HashMap<
-                sprout::tree::Root,
-                Arc<sprout::tree::NoteCommitmentTree>,
-            > = HashMap::new();
-
-            for joinsplit in transaction.sprout_groth16_joinsplits() {
-                // Check all anchor sets, including the one for interstitial
-                // anchors.
-                //
-                // The anchor is checked and the matching tree is obtained,
-                // which is used to create the interstitial tree state for this
-                // JoinSplit:
-                //
-                // > For each JoinSplit description in a transaction, an
-                // > interstitial output treestate is constructed which adds the
-                // > note commitments and nullifiers specified in that JoinSplit
-                // > description to the input treestate referred to by its
-                // > anchor. This interstitial output treestate is available for
-                // > use as the anchor of subsequent JoinSplit descriptions in
-                // > the same transaction.
-                //
-                // <https://zips.z.cash/protocol/protocol.pdf#joinsplit>
-                //
-                // # Consensus
-                //
-                // > The anchor of each JoinSplit description in a transaction
-                // > MUST refer to either some earlier block’s final Sprout
-                // > treestate, or to the interstitial output treestate of any
-                // > prior JoinSplit description in the same transaction.
-                //
-                // > For the first JoinSplit description of a transaction, the
-                // > anchor MUST be the output Sprout treestate of a previous
-                // > block.
-                //
-                // <https://zips.z.cash/protocol/protocol.pdf#joinsplit>
-                //
-                // Note that in order to satisfy the latter consensus rule above,
-                // [`interstitial_trees`] is always empty in the first iteration
-                // of the loop.
-                let input_tree = interstitial_trees
-                    .get(&joinsplit.anchor)
-                    .cloned()
-                    .or_else(|| {
-                        parent_chain
-                            .sprout_trees_by_anchor
-                            .get(&joinsplit.anchor)
-                            .cloned()
-                            .or_else(|| {
-                                finalized_state
-                                    .sprout_note_commitment_tree_by_anchor(&joinsplit.anchor)
-                            })
-                    });
+) -> Result<HashMap<sprout::tree::Root, Arc<sprout::tree::NoteCommitmentTree>>, ValidateContextError>
+{
+    let mut sprout_final_treestates = HashMap::new();
```

Comment:
I need to review this more carefully but on a quick glance I was kind of confused why the function was renamed and documentation changed to remove Sprout but it still does Sprout things

Reply:
I'll split this function further in PR #4802.